File: packages/core/src/routes/services.ts
import { Hono } from 'hono';
import { getDb, schema } from '../lib/db.js';
import { startService, stopService, restartService, getServiceLogs } from '../services/service-runner.js';

export const servicesRouter = new Hono();
const userIdFromCtx = () => 1;

servicesRouter.post('/', async (c) => {
  const db = await getDb();
  const user_id = userIdFromCtx();
  const body = await c.req.json().catch(() => ({} as any));
  const { name, type, port, entrypoint, working_dir, env_vars } = body;
  if (!name || !type) return c.json({ error: 'name and type required' }, 400);
  const [row] = await db
    .insert(schema.services)
    .values({ user_id, name, type, port: port ?? null, entrypoint: entrypoint || null, working_dir: working_dir || null, env_vars: env_vars || null, is_running: false } as any)
    .returning();
  return c.json(row);
});

servicesRouter.get('/', async (c) => {
  const db = await getDb();
  const user_id = userIdFromCtx();
  const rows = await db.select().from(schema.services).where({ user_id } as any);
  return c.json(rows);
});

servicesRouter.get('/:id', async (c) => {
  const db = await getDb();
  const id = Number.parseInt(c.req.param('id'), 10);
  const row = (await db.select().from(schema.services).where({ id } as any).limit(1))[0];
  if (!row) return c.json({ error: 'not found' }, 404);
  return c.json(row);
});

servicesRouter.patch('/:id', async (c) => {
  const db = await getDb();
  const id = Number.parseInt(c.req.param('id'), 10);
  const body = await c.req.json().catch(() => ({} as any));
  const updates: any = {};
  for (const k of ['name', 'type', 'port', 'entrypoint', 'working_dir', 'env_vars']) if (k in body) updates[k] = body[k];
  const [row] = await db.update(schema.services).set(updates).where({ id } as any).returning();
  return c.json(row);
});

servicesRouter.delete('/:id', async (c) => {
  const db = await getDb();
  const id = Number.parseInt(c.req.param('id'), 10);
  await stopService(id);
  await db.delete(schema.services).where({ id } as any);
  return c.json({ ok: true });
});

servicesRouter.post('/:id/start', async (c) => {
  const id = Number.parseInt(c.req.param('id'), 10);
  const status = await startService(id);
  return c.json({ ok: true, status });
});

servicesRouter.post('/:id/stop', async (c) => {
  const id = Number.parseInt(c.req.param('id'), 10);
  const status = await stopService(id);
  return c.json({ ok: true, status });
});

servicesRouter.post('/:id/restart', async (c) => {
  const id = Number.parseInt(c.req.param('id'), 10);
  const status = await restartService(id);
  return c.json({ ok: true, status });
});

servicesRouter.get('/:id/logs', async (c) => {
  const id = Number.parseInt(c.req.param('id'), 10);
  const lines = getServiceLogs(id, 1000);
  return c.json({ lines });
});

File: packages/api/app/routers/services.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.middleware.auth import get_current_user
from app.models.user import User
from app.schemas.services import ServiceCreate, ServiceUpdate
from app.services import service_manager

router = APIRouter(prefix="/api/services", tags=["services"])


@router.post("/")
async def create_service(body: ServiceCreate, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await service_manager.create_service(
        db, user.id, name=body.name, type=body.type,
        entrypoint=body.entrypoint, port=body.port, env_vars=body.env_vars,
    )


@router.get("/")
async def list_services(user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await service_manager.list_services(db, user.id)


@router.get("/{service_id}")
async def get_service(service_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    svc = await service_manager.get_service(db, user.id, service_id)
    if not svc:
        raise HTTPException(status_code=404, detail="Service not found")
    return svc


@router.patch("/{service_id}")
async def update_service(service_id: int, body: ServiceUpdate, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    svc = await service_manager.update_service(
        db, user.id, service_id,
        name=body.name, entrypoint=body.entrypoint, port=body.port, env_vars=body.env_vars,
    )
    if not svc:
        raise HTTPException(status_code=404, detail="Service not found")
    return svc


@router.delete("/{service_id}")
async def delete_service(service_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    deleted = await service_manager.delete_service(db, user.id, service_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Service not found")
    return {"ok": True}


@router.post("/{service_id}/start")
async def start_service(service_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    svc = await service_manager.get_service(db, user.id, service_id)
    if not svc:
        raise HTTPException(status_code=404, detail="Service not found")
    try:
        svc = await service_manager.start_service(db, svc)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"ok": True, "status": "started", "port": svc.port}


@router.post("/{service_id}/stop")
async def stop_service(service_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    svc = await service_manager.get_service(db, user.id, service_id)
    if not svc:
        raise HTTPException(status_code=404, detail="Service not found")
    await service_manager.stop_service(db, svc)
    return {"ok": True, "status": "stopped"}


@router.post("/{service_id}/restart")
async def restart_service(service_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    svc = await service_manager.get_service(db, user.id, service_id)
    if not svc:
        raise HTTPException(status_code=404, detail="Service not found")
    svc = await service_manager.restart_service(db, svc)
    return {"ok": True, "status": "restarted", "port": svc.port}


@router.get("/{service_id}/logs")
async def get_service_logs(service_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    svc = await service_manager.get_service(db, user.id, service_id)
    if not svc:
        raise HTTPException(status_code=404, detail="Service not found")
    lines = service_manager.get_logs(service_id)
    return {"lines": lines}
